### ISSUE TITLE
Messed up lines in `meta` file

### ISSUE THREAD

**Issue by matthiasgomolka** [ropensci/targets#368]
## Prework

* [x] Read and agree to the [code of conduct](https://ropensci.org/code-of-conduct/) and [contributing guidelines](https://github.com/ropensci/targets/blob/main/CONTRIBUTING.md).
* [x] Confirm that your issue is most likely a genuine bug in the `targets` package itself and not a user error, a known limitation, or a bug in another package that targets depends on. For miscellaneous troubleshooting, please post to the [discussions](https://github.com/ropensci/targets/discussions) instead.
* [x] If there is [already a relevant issue](https://github.com/ropensci/targets/issues), whether open or closed, comment on the existing thread instead of posting a new issue.
* [ ] Post a [minimal reproducible example](https://www.tidyverse.org/help/) like [this one](https://github.com/ropensci/targets/issues/256#issuecomment-754229683) so the maintainer can troubleshoot the problems you identify. A reproducible example is:
    * [ ] **Runnable**: post enough R code and data so any onlooker can create the error on their own computer.
    * [ ] **Minimal**: reduce runtime wherever possible and remove complicated details that are irrelevant to the issue at hand.
    * [ ] **Readable**: format your code according to the [tidyverse style guide](https://style.tidyverse.org/).

## Description

When I run larger plans, some lines of my `meta` file are be messed up (at random, as far as I could observe):

```
...
files_37eacef2|branch|26cfdb6d8f8154f2|1c796631dd062329|0a758515684cdf8c|1122515995|_targets/objects/files_37eacef2|d1525e8e93a90b4e|6b4bfcd70e6e5a55|207|rds|vector|files||0||

les_242150f9|branch|b96487a841b775aa|1c796631dd062329|9268899a6c0fa97b|779605311|_targets/objects/files_242150f9|fbb9fc530f9f345a|6b4bfcd70e6e5a55|207|rds|vector|files||0.02||files_edb565bb|branch|bab54f0432333572|1c796631dd062329|dcb7453b42e6e5bd|366967898|_targets/objects/files_edb565bb|b7f9cb03da0353cd|6b4bfcd70e6e5a55|207|rds|vector|files||0||
files_ea41d834|branch|30abb6a07877ffa0|1c796631dd062329|2cb49c8c8912640c|-2040554421|_targets/objects/files_ea41d834|dcb8983fd832f683|6b4bfcd70e6e5a55|207|rds|vector|files||0||
...
```

I notice this for example when running `tar_visnetwork()`:

```r
> targets::tar_visnetwork()
Warning messages:
1: package 'targets' was built under R version 4.0.4 
2: package 'tarchetypes' was built under R version 4.0.4 
3: package 'future' was built under R version 4.0.3 
4: package 'visNetwork' was built under R version 4.0.3 
5: package 'fst' was built under R version 4.0.3 
6: In data.table::fread(file = self$path, sep = database_sep_outer,  :
  Stopped early on line 2734. Expected 17 fields but found 33. Consider fill=TRUE and comment.char=. First discarded non-empty line: <<les_242150f9|branch|b96487a841b775aa|1c796631dd062329|9268899a6c0fa97b|779605311|_targets/objects/files_242150f9|fbb9fc530f9f345a|6b4bfcd70e6e5a55|207|rds|vector|files||0.02||files_edb565bb|branch|bab54f0432333572|1c796631dd062329|dcb7453b42e6e5bd|366967898|_targets/objects/files_edb565bb|b7f9cb03da0353cd|6b4bfcd70e6e5a55|207|rds|vector|files||0||>>
7: In data.table::fread(file = self$path, sep = database_sep_outer,  :
  Stopped early on line 618. Expected 5 fields but found 9. Consider fill=TRUE and comment.char=. First discarded non-empty line: <<les_0c86d3ee|branch|files|0|startedfiles_0c86d3ee|branch|files|0|built>>
```

When I repair the file manually (only a few lines are messed up) like this ...

```
files_37eacef2|branch|26cfdb6d8f8154f2|1c796631dd062329|0a758515684cdf8c|1122515995|_targets/objects/files_37eacef2|d1525e8e93a90b4e|6b4bfcd70e6e5a55|207|rds|vector|files||0||
files_242150f9|branch|b96487a841b775aa|1c796631dd062329|9268899a6c0fa97b|779605311|_targets/objects/files_242150f9|fbb9fc530f9f345a|6b4bfcd70e6e5a55|207|rds|vector|files||0.02||
files_edb565bb|branch|bab54f0432333572|1c796631dd062329|dcb7453b42e6e5bd|366967898|_targets/objects/files_edb565bb|b7f9cb03da0353cd|6b4bfcd70e6e5a55|207|rds|vector|files||0||
files_ea41d834|branch|30abb6a07877ffa0|1c796631dd062329|2cb49c8c8912640c|-2040554421|_targets/objects/files_ea41d834|dcb8983fd832f683|6b4bfcd70e6e5a55|207|rds|vector|files||0||
```

... `tar_visnetwork()` runs without errors.

When I rerun the plan, this triggers the dynamic target to be outdated (which is understandable because not all subtargets have a correct entry in `meta`).

This reminds me of https://github.com/ropensci/drake/issues/1331, where I remember that the issue was only reduced, but not completely solved after the fix.

## Reproducible example

I did not manage to build a reprex, since this seems to happen at random. I did not observe this in small examples.

## Expected result

The `meta` file should be written correctly.

## Diagnostic information

```r
> sessionInfo()
R version 4.0.2 (2020-06-22)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 17763)

Matrix products: default

locale:
[1] LC_COLLATE=German_Germany.1252  LC_CTYPE=German_Germany.1252    LC_MONETARY=German_Germany.1252 LC_NUMERIC=C                   
[5] LC_TIME=German_Germany.1252    

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
[1] sdcLog_0.2.0      arrow_3.0.0       data.table_1.14.0 fs_1.5.0          fst_0.9.4         visNetwork_2.0.9  future_1.19.1    
[8] tarchetypes_0.1.0 targets_0.2.0    

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5        pillar_1.4.7      compiler_4.0.2    tools_4.0.2       bit_4.0.4         digest_0.6.27     checkmate_2.0.0  
 [8] jsonlite_1.7.1    lifecycle_0.2.0   tibble_3.0.4      pkgconfig_2.0.3   rlang_0.4.9       igraph_1.2.6      cli_2.2.0        
[15] rstudioapi_0.13   yaml_2.2.1        parallel_4.0.2    dplyr_1.0.2       withr_2.3.0       generics_0.1.0    vctrs_0.3.5      
[22] htmlwidgets_1.5.2 globals_0.13.1    bit64_4.0.5       tidyselect_1.1.0  glue_1.4.2        listenv_0.8.0     R6_2.5.0         
[29] processx_3.4.4    fansi_0.4.1       tidyr_1.1.2       callr_3.5.1       purrr_0.3.4       magrittr_2.0.1    backports_1.2.0  
[36] codetools_0.2-16  ps_1.4.0          ellipsis_0.3.1    htmltools_0.5.0   assertthat_0.2.1  renv_0.12.5       broom_0.7.1      
[43] crayon_1.3.4
```

**Comment by wlandau** [comment]
With https://github.com/ropensci/drake/issues/1331, multiple parallel workers were writing to the same log, and the issue went away when this was restricted to only the main process. In `targets`, only the main process writes to the metadata, so I am surprised to hear of such an issue.

Are you running multiple instances of `tar_make()` concurrently? At most one run of the pipeline should be active at any given time. I thought about implementing a guardrail for this but decided against it. (`drake` has this guardrail but it mostly causes annoying false positives.)

Alternatively, how exactly are you invoking the pipeline? Are you using parallel workers with `tar_make_clustermq()` or `tar_make_future()`. This probably won't matter, but in case I am wrong about the multiple-`tar_make()` theory, this is another thing to look at.

**Comment by matthiasgomolka** [comment]
This happens with `tar_make_future()` and with `tar_make()`. And it happens quite often and always in the same way (as you can see in the example):

- One blank line
- the next line is missing the first two characters
- and it contains the next line as well

I'm only running one `tar_make()` at a time. Do other functions such as `tar_read()` also count here?

**Comment by matthiasgomolka** [comment]
I'm working on a reprex. I hope I can somehow reproduce the problem consistently.

**Comment by wlandau** [comment]
I wonder if this is another system-specific Windows network drive issue. What happens if you try this?

``` r
line <- "files_37eacef2|branch|26cfdb6d8f8154f2|1c796631dd062329|0a758515684cdf8c|1122515995|_targets/objects/files_37eacef2|d1525e8e93a90b4e|6b4bfcd70e6e5a55|207|rds|vector|files||0||"
path <- tempfile() # Change this to a file on a slow network drive.
tmp <- replicate(1e4, write(line, path, ncolumns = 1L, append = TRUE, sep = ""))
lines <- readLines(path)
table(vapply(lines, nchar, FUN.VALUE = integer(1)))
#> 
#>   175 
#> 10000
```

<sup>Created on 2021-03-25 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by matthiasgomolka** [comment]
> I wonder if this is another system-specific Windows network drive issue. What happens if you try this?
> 
> ```r
> line <- "files_37eacef2|branch|26cfdb6d8f8154f2|1c796631dd062329|0a758515684cdf8c|1122515995|_targets/objects/files_37eacef2|d1525e8e93a90b4e|6b4bfcd70e6e5a55|207|rds|vector|files||0||"
> path <- tempfile() # Change this to a file on a slow network drive.
> tmp <- replicate(1e4, write(line, path, ncolumns = 1L, append = TRUE, sep = ""))
> lines <- readLines(path)
> table(vapply(lines, nchar, FUN.VALUE = integer(1)))
> #> 
> #>   175 
> #> 10000
> ```
> 
> Created on 2021-03-25 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)

Yes, you're right! This also produces the exact same error, already at line 27 and pretty often afterwards:

```r
> line <- "files_37eacef2|branch|26cfdb6d8f8154f2|1c796631dd062329|0a758515684cdf8c|1122515995|_targets/objects/files_37eacef2|d1525e8e93a90b4e|6b4bfcd70e6e5a55|207|rds|vector|files||0||"
> path <- "path_on_slow_network_drive"
> tmp <- replicate(1e4, write(line, path, ncolumns = 1L, append = TRUE, sep = ""))
> lines <- readLines(path)
> table(vapply(lines, nchar, FUN.VALUE = integer(1)))

   0  173  175  348 
 232   32 9504  232 
> 
```

I should really try get a faster network drive connection.

Anyways, do you think there is anything `targets` can do about this?

**Comment by wlandau** [comment]
Thanks for confirming. As long as all the metadata lives in the same file (an important design choice that resolves egregious performance issues that were in `drake`) there is nothing `targets` can do.

Would you be willing to follow up with this on Stack Overflow and/or the R-devel mailing list? Latency should not create a race condition for a single-threaded task.

**Comment by matthiasgomolka** [comment]
> Would you be willing to follow up with this on Stack Overflow and/or the R-devel mailing list? Latency should not create a race condition for a single-threaded task.

Yes, I will. If anything comes up, I'll post the link here for future reference.

Thanks for tracking down the underlying issue!

**Comment by KGallagher7** [comment]
Hi all, 

I am also having an issue with my meta file on a Windows network drive. I am working out of a network folder mounted into the RStudio container using SMB/CIFS. My issue comes from not the meta file having errors, but the meta file is being deleted at the end of my workflow.  

Here you can see that the workflow completes without issue on start up: 

<img width="856" height="366" alt="Image" src="https://github.com/user-attachments/assets/5d0dcaf4-d37f-4d0a-b4ba-07f253ed6c39" />
(The warnings are from the function called and can be ignored) 

I then try to add a few more steps to the workflow, and the workflow completes, but produces a warning that the meta file does not exist.

<img width="874" height="139" alt="Image" src="https://github.com/user-attachments/assets/6982e129-931a-4373-b2b4-f321abc93381" />

The meta file was present at the end of the first iteration of the run, but after adding these few extra targets, it is gone. 

I worked with my IT department to look on the backend of the container and there doesn't appear to be any CPU or RAM limitations. They also saw the meta file as it was being written and everything looked fine, but saw no commands that would delete the file. Permissions are wide open in the directory. Originally, I was only getting these errors when working in parallel using crew, but recently started experiencing them working sequentially as well. Creating fresh workflows with tar_destroy() has not worked.

I've been able to get this same workflow working locally on my PC (both in parallel and sequentially), as well as in my container in a non-network mounted folder, but these folders are cleared out every time the container is launched, so I believe that this is a bug within targets not working with the network-mounted folder. Was this issue with targets/R not working well with network drives ever resolved? 

Thanks!
Katherine 

PS - If I should start a new issue, please let me know. It may be hard to provide a reproducible example given the issue, but happy to do so if that would be helpful!
